File: Java Projects/Coffee Machine (Easy)/Stage 6/CoffeeMachine.java
package machine;

import java.util.Scanner;

public class CoffeeMachine {
    public static void main(String[] args) {
        while (true) {
            if (state == ValidStates.IDLE) {
                System.out.println("Write action (buy, fill, take, remaining, exit):");
                state = ValidStates.PROCESSING;
            } else if (state != ValidStates.OFF) {
                processInput(scanner.next());
            } else if (state == ValidStates.OFF) {
                break;
            }
        }
    }

    public static void processInput(String userInput) {
        if (state == ValidStates.PROCESSING) {
            System.out.println();
            switch (userInput) {
                case "buy":
                    System.out.println("What do you want to buy? 1 - espresso, " +
                            "2 - latte, 3 - cappuccino, back - to main menu: ");
                    state = ValidStates.BUYING;
                    break;
                case "fill":
                    state = ValidStates.FILLING;
                    fill(-1);
                    break;
                case "take":
                    take();
                    state = ValidStates.IDLE;
                    System.out.println();
                    break;
                case "remaining":
                    remaining();
                    state = ValidStates.IDLE;
                    System.out.println();
                    break;
                case "exit":
                    state = ValidStates.OFF;
                    return;
            }
        } else if (state == ValidStates.FILLING && fillState != FillStates.NA) {
            fill(Integer.parseInt(userInput));
            if (fillState == FillStates.NA) {
                state = ValidStates.IDLE;
                System.out.println();
            }
        } else if (state == ValidStates.BUYING) {
            buy(userInput);
            state = ValidStates.IDLE;
            System.out.println();
        }
    }
    
    private static void remaining() {
        String[] identifier = {
                "water",
                "milk",
                "coffee beans",
                "disposable cups",
                "money"
        };
        System.out.println("The coffee machine has:");
        for (int i = 0; i < 5; i++) {
            if (i == 4) {
                System.out.print('$');
            }
            System.out.println(inventory[i] + " of " + identifier[i]);
        }
    }

    private static void take() {
        System.out.println("I give you " + inventory[4]);
        adjustInventory(0, 0, 0, 0, -inventory[4]);
    }

    private static void fill(int input) {
        switch (fillState) {
            case NA:
                System.out.println("Write how many ml of water do you want to add:");
                fillState = FillStates.WATER;
                break;
            case WATER:
                inventory[0] += input;
                System.out.println("Write how many ml of milk do you want to add:");
                fillState = FillStates.MILK;
                break;
            case MILK:
                inventory[1] += input;
                System.out.println("Write how many grams of coffee beans do you want to add:");
                fillState = FillStates.BEANS;
                break;
            case BEANS:
                inventory[2] += input;
                System.out.println("Write how many disposable cups of coffee do you want to add:");
                fillState = FillStates.CUPS;
                break;
            case CUPS:
                inventory[3] += input;
                fillState = FillStates.NA;
        }
    }

    private static void buy(String choice) {
        boolean success = false;
        switch (choice) {
            case "1":
                success = buyEspresso();
                break;
            case "2":
                success = buyLatte();
                break;
            case "3":
                success = buyCappuccino();
                break;
            case "back":
                return;
        }
        if (success) {
            System.out.println("I have enough resources, making you a coffee!");
        }
    }

    private static boolean buyEspresso() {
        if (checkResources(250, 0, 16)) {
            adjustInventory(-250, 0, -16, -1, 4);
            return true;
        } else {
            return false;
        }
    }

    private static boolean buyLatte() {
        if (checkResources(350, 75, 20))
        {
            adjustInventory(-350, -75, -20, -1, 7);
            return true;
        } else {
            return false;
        }
    }

    private static boolean buyCappuccino() {
        if (checkResources(200, 100, 12)) {
            adjustInventory(-200, -100, -12, -1, 6);
            return true;
        } else {
            return false;
        }
    }

    private static boolean checkResources(int water, int milk, int beans) {
        if (water > inventory[0]) {
            System.out.println("Sorry, not enough water!");
            return false;
        } else if (milk > inventory[1]) {
            System.out.println("Sorry, not enough milk!");
            return false;
        } else if (beans > inventory[2]) {
            System.out.println("Sorry, not enough coffee beans!");
            return false;
        } else if (1 > inventory[3]) {
            System.out.println("Sorry, not enough disposable cups!");
            return false;
        }
        return true;
    }

    private static void adjustInventory(int water, int milk, int beans, int cups, int cash) {
        inventory[0] += water;
        inventory[1] += milk;
        inventory[2] += beans;
        inventory[3] += cups;
        inventory[4] += cash;
    }

    private enum ValidStates {
        IDLE, FILLING, BUYING, PROCESSING, OFF
    }

    private enum FillStates {
        NA, WATER, MILK, BEANS, CUPS
    }

    private static ValidStates state = ValidStates.IDLE;
    private static FillStates fillState = FillStates.NA;
    private static Scanner scanner = new Scanner(System.in);
    private static int[] inventory = {400, 540, 120, 9, 550};
}

File: Python Projects/Coffee Machine (Easy)/Stage 5/coffee_machine.py
available_ml_water = 400
available_ml_milk = 540
available_g_beans = 120
available_cups = 9
available_cash = 550


def buy():
    global available_cups
    print("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, back - to main menu: ")
    choice = input()
    if choice == "back":
        return
    else:
        if available_cups == 0:
            print("Sorry, not enough cups!")
            return
        if choice == "1":
            coffee_created = create_espresso()
        elif choice == "2":
            coffee_created = create_latte()
        elif choice == "3":
            coffee_created = create_cappuccino()
        if coffee_created:
            available_cups -= 1
            print("I have enough resources, making you a coffee!")


def create_espresso():
    global available_ml_water
    global available_g_beans
    global available_cash
    if available_ml_water < 250:
        print("Sorry, not enough water!")
        return False
    elif available_g_beans < 16:
        print("Sorry, not enough coffee beans!")
        return False
    else:
        available_cash += 4
        available_ml_water -= 250
        available_g_beans -= 16
        return True


def create_latte():
    global available_ml_water
    global available_ml_milk
    global available_g_beans
    global available_cash
    if available_ml_water < 350:
        print("Sorry, not enough water!")
        return False
    elif available_ml_milk < 75:
        print("Sorry, not enough milk!")
        return False
    elif available_g_beans < 20:
        print("Sorry, not enough coffee beans!")
        return False
    else:
        available_cash += 7
        available_ml_water -= 350
        available_ml_milk -= 75
        available_g_beans -= 20
        return True


def create_cappuccino():
    global available_ml_water
    global available_ml_milk
    global available_g_beans
    global available_cash
    if available_ml_water < 200:
        print("Sorry, not enough water!")
        return False
    elif available_ml_milk < 100:
        print("Sorry, not enough milk!")
        return False
    elif available_g_beans < 12:
        print("Sorry, not enough coffee beans!")
        return False
    else:
        available_cash += 6
        available_ml_water -= 200
        available_ml_milk -= 100
        available_g_beans -= 12
        return True


def fill_supplies():
    global available_ml_water
    global available_ml_milk
    global available_g_beans
    global available_cups

    print("Write how many ml of water do you want to add:")
    available_ml_water += int(input())
    print("Write how many ml of milk do you want to add:")
    available_ml_milk += int(input())
    print("Write how many grams of coffee beans do you want to add:")
    available_g_beans += int(input())
    print("Write how many disposable cups of coffee do you want to add:")
    available_cups += int(input())


def take():
    global available_cash
    temp = available_cash
    available_cash = 0
    print(f"I gave you ${temp}")
    return temp


def print_status():
    global available_ml_water
    global available_ml_milk
    global available_g_beans
    global available_cups
    global available_cash
    print("The coffee machine has: ")
    print(f"{available_ml_water} of water")
    print(f"{available_ml_milk} of milk")
    print(f"{available_g_beans} of coffee beans")
    print(f"{available_cups} of disposable cups")
    if available_cash > 0:
        print(f"${available_cash} of money")
    else:
        print("0 of money")


def run():
    while True:
        print("Write action (buy, fill, take, remaining, exit):")
        choice = input()
        print()
        if choice == "buy":
            buy()
        elif choice == "fill":
            fill_supplies()
        elif choice == "take":
            take()
        elif choice == "remaining":
            print_status()
        elif choice == "exit":
            break
        print()


run()
